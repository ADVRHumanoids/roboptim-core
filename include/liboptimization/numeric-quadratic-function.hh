// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericQuadraticFunction class.
 */

#ifndef OPTIMIZATION_NUMERIC_QUADRATIC_FUNCTION_HH
# define OPTIMIZATION_NUMERIC_QUADRATIC_FUNCTION_HH
# include <boost/numeric/ublas/symmetric.hpp>

# include <liboptimization/quadratic-function.hh>

namespace optimization
{
  namespace ublas = boost::numeric::ublas;

  /// Define a quadratic function.
  class NumericQuadraticFunction : public QuadraticFunction
  {
  public:
    NumericQuadraticFunction (const matrix_t&, const vector_t&)
      throw ();

    virtual vector_t operator () (const vector_t&) const throw ();
    virtual gradient_t gradient (const vector_t&, int) const throw ();
    virtual hessian_t hessian (const vector_t&, int) const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  private:
    ublas::symmetric_matrix<double, ublas::lower> a_;
    vector_t b_;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_QUADRATIC_FUNCTION_HH
