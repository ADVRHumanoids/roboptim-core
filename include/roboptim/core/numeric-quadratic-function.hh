// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the NumericQuadraticFunction class.
 */

#ifndef ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HH
# define ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HH
# include <boost/numeric/ublas/symmetric.hpp>

# include <roboptim/core/quadratic-function.hh>

namespace roboptim
{
  /**
     \addtogroup roboptim_function
     @{
  */
  namespace ublas = boost::numeric::ublas;

  /// Define a quadratic function.
  class NumericQuadraticFunction : public QuadraticFunction
  {
  public:
    NumericQuadraticFunction (const matrix_t&, const vector_t&)
      throw ();

    ~NumericQuadraticFunction () throw ();

    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, int = 0) const throw ();
    void impl_hessian (hessian_t& hessian,
		       const argument_t& argument,
		       int functionId = 0) const throw ();
  private:
    ublas::symmetric_matrix<double, ublas::lower> a_;
    vector_t b_;
  };
/**
   @}
*/

} // end of namespace roboptim
#endif //! ROBOPTIM_CORE_QUADRATIC_FUNCTION_HH
