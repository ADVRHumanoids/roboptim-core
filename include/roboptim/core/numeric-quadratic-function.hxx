// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HXX
# define ROBOPTIM_CORE_NUMERIC_QUADRATIC_FUNCTION_HXX
# include "debug.hh"

# include <roboptim/core/indent.hh>
# include <roboptim/core/numeric-linear-function.hh>
# include <roboptim/core/util.hh>

namespace roboptim
{
  template <typename T>
  GenericNumericQuadraticFunction<T>::GenericNumericQuadraticFunction
  (const matrix_t& a, const vector_t& b)
    throw ()
    : GenericQuadraticFunction<T>
      (a.rows (), 1, "numeric linear function"),
      a_ (a),
      b_ (b),
      buffer_ (b.size ())
  {
    assert (b.size () == this->inputSize ());
  }

  template <typename T>
  GenericNumericQuadraticFunction<T>::~GenericNumericQuadraticFunction () throw ()
  {
  }

  // 1/2 * x^T * A * x + b^T * x
  template <typename T>
  void
  GenericNumericQuadraticFunction<T>::impl_compute (result_t& result,
						 const argument_t& argument)
    const throw ()
  {
    buffer_.noalias () = a_ * argument;
    result = .5 * argument.adjoint ()  * buffer_;
    result += b_.adjoint () * argument;
  }

  // x * A + b
  template <>
  inline void
  GenericNumericQuadraticFunction<EigenMatrixSparse>::impl_jacobian
  (jacobian_t& jacobian, const argument_t& x) const throw ()
  {
    jacobian = x.transpose () * a_.selfadjointView<Eigen::Upper> ();
    for (size_type j = 0; j < this->inputSize (); ++j)
      jacobian.coeffRef (0, j) += b_[j];
  }


  template <typename T>
  void
  GenericNumericQuadraticFunction<T>::impl_jacobian
  (jacobian_t& jacobian, const argument_t& x) const throw ()
  {
    jacobian.noalias () = x.transpose () * a_;
    jacobian += b_.transpose ();
  }

  // A(i) - sparse specialization
  template <>
  inline void
  GenericNumericQuadraticFunction<EigenMatrixSparse>::impl_gradient
  (gradient_t& gradient, const argument_t& x, size_type)
    const throw ()
  {
    buffer_.noalias () = a_ * x;
    buffer_ += b_;
    for (size_type j = 0; j < this->inputSize (); ++j)
      gradient.coeffRef (j) = buffer_.coeffRef (j);
  }

  // A(i)
  template <typename T>
  void
  GenericNumericQuadraticFunction<T>::impl_gradient
  (gradient_t& gradient, const argument_t& x, size_type) const throw ()
  {
    gradient.noalias () = a_ * x;
    gradient += b_;
  }

  // A
  template <typename T>
  void
  GenericNumericQuadraticFunction<T>::impl_hessian
  (hessian_t& hessian, const argument_t&, size_type) const throw ()
  {
    hessian = a_;
  }

  template <typename T>
  std::ostream&
  GenericNumericQuadraticFunction<T>::print (std::ostream& o) const throw ()
  {
    return o << "Numeric quadratic function" << incindent << iendl
             << "A = " << this->a_ << iendl
             << "B = " << this->b_
             << decindent;
  }

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_QUADRATIC_FUNCTION_HXX
