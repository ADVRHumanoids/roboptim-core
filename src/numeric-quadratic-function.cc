// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <boost/numeric/ublas/io.hpp>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/numeric/ublas/vector.hpp>


#include <roboptim/core/indent.hh>
#include <roboptim/core/numeric-quadratic-function.hh>

namespace roboptim
{
  NumericQuadraticFunction::NumericQuadraticFunction (const symmetric_t& a,
                                                      const vector_t& b)
    throw ()
    : QuadraticFunction (a.size1 (), 1),
      a_ (a),
      b_ (b)
  {
    assert (a.size1 () == a.size2 () && a.size2 () == b.size ());
  }


  NumericQuadraticFunction::~NumericQuadraticFunction () throw ()
  {
  }


  // 1/2 * x^T * A * x + b^T * x
  void
  NumericQuadraticFunction::impl_compute (result_t& result,
					  const argument_t& argument)
    const throw ()
  {
    using namespace boost::numeric::ublas;

    result (0) = inner_prod
      (prod (trans (argument), a_), argument) / 2
      + inner_prod (trans (b_), argument);
  }

  // x * A + b
  void
  NumericQuadraticFunction::impl_gradient (result_t& result,
					   const argument_t& x,
					   size_type) const throw ()
  {
    using namespace boost::numeric::ublas;
    noalias (result) = prod (x, a_) + b_;
  }

  // A
  void
  NumericQuadraticFunction::impl_hessian (hessian_t& hessian,
					  const argument_t&,
					  size_type) const throw ()
  {
    hessian = a_;
  }

  std::ostream&
  NumericQuadraticFunction::print (std::ostream& o) const throw ()
  {
    return o << "Numeric quadratic function" << incindent << iendl
             << "A = " << a_ << iendl
             << "B = " << b_
             << decindent;
  }

} // end of namespace roboptim
