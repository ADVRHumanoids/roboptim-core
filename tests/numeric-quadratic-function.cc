// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/numeric-quadratic-function.hh>

#include "common.hh"

using namespace roboptim;

typedef DummySolver solver_t;

int run_test ()
{
  NumericQuadraticFunction::matrix_t a (5, 5);
  NumericQuadraticFunction::vector_t b (5);
  NumericQuadraticFunction::vector_t x (5);

  a.clear ();
  b.clear ();
  x.clear ();

  a(0, 0) = 1.1, a(0, 1) = 1.2, a(0, 2) = 1.3, a(0, 3) = 1.4, a(0, 4) = 1.5;
  a(1, 0) = 1.2, a(1, 1) = 2.2, a(1, 2) = 2.3, a(1, 3) = 2.4, a(1, 4) = 2.5;
  a(2, 0) = 1.3, a(2, 1) = 2.3, a(2, 2) = 3.3, a(2, 3) = 3.4, a(2, 4) = 3.5;
  a(3, 0) = 1.4, a(3, 1) = 2.4, a(3, 2) = 3.4, a(3, 3) = 4.4, a(3, 4) = 4.5;
  a(4, 0) = 1.5, a(4, 1) = 2.5, a(4, 2) = 3.5, a(4, 3) = 4.5, a(4, 4) = 5.5;

  b[0] = 2.1;
  b[1] = 4.3;
  b[2] = 6.5;
  b[3] = 8.7;

  NumericQuadraticFunction f (a, b);

  std::cout << f << std::endl;

  x[0] = 0.1;
  x[1] = 1.2;
  x[2] = 2.3;
  x[3] = 3.4;
  x[4] = 4.5;

  std::cout << "f(x) = " << f (x) << std::endl;
  std::cout << "J(x) = " << f.jacobian (x) << std::endl;
  std::cout << "G(x) = " << f.gradient (x, 0) << std::endl;
  std::cout << "H(x) = " << f.hessian (x, 0) << std::endl;

  return 0;
}

GENERATE_TEST ()
