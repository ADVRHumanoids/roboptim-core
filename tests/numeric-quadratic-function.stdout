Dummy (numeric quadratic function):
  A = [5,5]((1,0,0,0,0), (0,1,0,0,0), (0,0,1,0,0), (0,0,0,1,0), (0,0,0,0,1))
  B = [5](0,0,0,0,0)
  c = [1](0)
x = [5](0,0,0,0,0)
f(x) = [1](0)
J(x) = [5](0,0,0,0,0)
G(x) = [5](0,0,0,0,0)
H(x) = [5,5]((1,0,0,0,0), (0,1,0,0,0), (0,0,1,0,0), (0,0,0,1,0), (0,0,0,0,1))
